export type REG_B = "REG_B"
declare REG_B: REG_B --# 1
export type REG_C = "REG_C"
declare REG_C: REG_C --# 2
export type REG_D = "REG_D"
declare REG_D: REG_D --# 3
export type REG_E = "REG_E"
declare REG_E: REG_E --# 4
export type REG_H = "REG_H"
declare REG_H: REG_H --# 5
export type REG_L = "REG_L"
declare REG_L: REG_L --# 6
export type REG_HLA = "REG_HLA"
declare REG_HLA: REG_HLA --# 7
export type REG_A = "REG_A"
declare REG_A: REG_A --# 8
export type REG_F = "REG_F"
declare REG_F: REG_F --# 9
export type REG_PC = "REG_PC"
declare REG_PC: REG_PC --# 10
export type REG_SP = "REG_SP"
declare REG_SP: REG_SP --# 11
export type REG_STOP = "REG_STOP"
declare REG_STOP: REG_STOP --# 12
export type REG_INTERRUPTS = "REG_INTERRUPTS"
declare REG_INTERRUPTS: REG_INTERRUPTS --# 13
export type MEMORY_ROM = "MEMORY_ROM"
declare MEMORY_ROM: MEMORY_ROM --# 14
export type MEMORY_WRAM = "MEMORY_WRAM"
declare MEMORY_WRAM: MEMORY_WRAM --# 15
export type MEMORY_ERAM = "MEMORY_ERAM"
declare MEMORY_ERAM: MEMORY_ERAM --# 16

export type REG_IDX =
	  REG_B
	| REG_C
	| REG_D
	| REG_E
	| REG_H
	| REG_L
	| REG_HLA
	| REG_A
	| REG_F
	| REG_PC
	| REG_SP
